rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+implementing extensive pieces of infrastructure, but at the same time provide
+the most value to the TF Lite team.
+
+We analysed how many kernels we can support in the future and include our
+findings in the following spreadsheets. As we describe in
+[Design Proposal](#design-proposal) below, Kernel Fallback depends on
+customizing
+[OpKernelConstruction](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=256?q=OpKernelConstruction)
+and
+[OpKernelContext](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=584?q=OpKernelContext&ss=tensorflow%2Ftensorflow)
+classes. Number of supported kernels will depend on the surface we manage to
+customize. (Note that I have already started prototyping the implementation that
+includes a few common methods such as `input`, `output`. The spreadsheet below
+consideres these methods to be already *supported*).
+
+*   List of kernels and `OpKernelConstruction`/`OpKernelContext` methods they
+    require:
+    [here](https://github.com/annarev/community/blob/tfrt_kernel_fallback_rfc/rfcs/20200712-tfrt-kernel-fallback/support_priority.md)
+*   Proposed implementation order for these methods:
+    [here](https://github.com/annarev/community/blob/tfrt_kernel_fallback_rfc/rfcs/20200712-tfrt-kernel-fallback/kernel_to_unsupported.md)
+
+Based on these estimates, we can support >= 423 kernels. Note that this number
+is just based on the `OpKernelConstruction`/`OpKernelContext` coverage that we
+can provide. It doesn't take into consideration other issues we might face.
+
+### TFRT Integration Goals
+
+We want to support executing a [BEF](https://github.com/tensorflow/runtime/blob/master/documents/binary_executable_format.md ) file
+on mobile device that calls kernels using Kernel Fallback mechanism. Users will
+be able to generate a BEF file based on a saved model and we will provide a
+script to create it.
+
+We might also want to support running ops using TFRT eager mode (that is, add a
+custom
+[OpHandler](https://github.com/tensorflow/runtime/blob/3c7a1ea02c87325f1b47aebb24b3ca6e84e7e7e7/include/tfrt/core_runtime/op_handler.h#L47)).
+
+## Non-goals
+
+*   Supporting all existing ops. `OpKernelContext` surface is quite large and
+    implementing all of it would require a significant amount of time. Instead,
+    we will start by adding most common and easy functionality. If certain
+    functionality is only used by a handful of kernels, it might make sense to
+    implement TFRT native kernels or rely on runtime fallback instead. One notable example is
+    [ResourceMgr](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/resource_mgr.h;l=152?q=ResourceMgr).
+    We might support it later, but it is definitely not first priority due to
+    extra effort required.
+*   Gradients would not be supported by the first iteration of Kernel Fallback,
```

Comment:
can you elaborate on what this means? IIUC gradient kernels are just like other normal kernels, but there is extra metadata that associate them with the forward-pass ops/kernels, which the fallback mechanism need not care.

So I imagine there could be gradient kernels that are also "well behaved" for kernel fallback purposes.

Reply:
Basically, I didn't look into what it will take to support gradients and decided to postpone this investigation till later.

Based on further discussion, seems like gradient kernels are just normal kernels but there might be further work in registering and calling them. For now, I am just keeping wording in this doc the way it is. We can revisit gradients later.